File: Leet/3.longest_substring_without_repeating_characters-最长无重复子串长度/3.longest_substring_without_repeating_characters.js
/**
 * @param {string} s
 * @return {number}
 */
 var lengthOfLongestSubstring = function(s) {
     let hm = new Map();
     let max = 0; let start = 0;

     for (let end = 0; end < s.length; end++) {
         let ch = s.charAt(end);
         if (hm.has(ch)) {
             start = Math.max(hm.get(ch) + 1, start);
         }
         max = Math.max(max, end - start + 1);
         hm.set(ch, end);
     }
     return max;
};

File: Algorithms/3.longest_substring_without_repeating_characters-最长无重复子串长度/3.longest_substring_without_repeating_characters.java
import java.util.HashMap;

class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> hm = new HashMap<>();
        int max = 0; int start = 0;

        for (int end = 0; end < s.length(); end++) {
            char ch = s.charAt(end);
            if (hm.containsKey(ch)) {
                start = Math.max(hm.get(ch) + 1, start);
            }
            max = Math.max(max, (end - start + 1));
            hm.put(ch, end);
        }
        return max;
    }
}